bookkeeper-server/src/main/java/org/apache/bookkeeper/client/ReadOnlyLedgerHandle.java

```diff
@@ -207,4 +206,197 @@ public void readLastConfirmedComplete(int rc, long lastConfirmed, Object ctx) {
+        }
+        new MetadataUpdateLoop(
+                clientCtx.getLedgerManager(), getId(),
+                this::getLedgerMetadata,
+                needsUpdate,
+                (metadata) -> LedgerMetadataBuilder.from(metadata).withInRecoveryState().build(),
+                this::setLedgerMetadata)
+            .run()
+            .thenCompose((metadata) -> {
+                    if (metadata.isClosed()) {
+                        return CompletableFuture.<Void>completedFuture(null);
+                    } else {
+                        return new LedgerRecoveryOp(ReadOnlyLedgerHandle.this, clientCtx)
+                            .setEntryListener(listener)
+                            .initiate();
+                    }
+            })
+            .thenCompose((ignore) -> closeRecovered())
+            .whenComplete((ignore, ex) -> {
+                    if (ex != null) {
+                        cb.operationComplete(getExceptionCode(ex, BKException.Code.UnexpectedConditionException), null);
+                    } else {
+                        cb.operationComplete(BKException.Code.OK, null);
+                    }
+            });
+    }
+
+    CompletableFuture<LedgerMetadata> closeRecovered() {
+        long lac, len;
+        synchronized (this) {
+            lac = lastAddConfirmed;
+            len = length;
+        }
+        LOG.info("Closing recovered ledger {} at entry {}", getId(), lac);
+        CompletableFuture<LedgerMetadata> f = new MetadataUpdateLoop(
+                clientCtx.getLedgerManager(), getId(),
+                this::getLedgerMetadata,
+                (metadata) -> metadata.isInRecovery(),
+                (metadata) -> {
+                    LedgerMetadataBuilder builder = LedgerMetadataBuilder.from(metadata);
+                    Optional<Long> lastEnsembleKey = metadata.getLastEnsembleKey();
+                    checkState(lastEnsembleKey.isPresent(),
+                               "Metadata shouldn't have been created without at least one ensemble");
+                    newEnsemblesFromRecovery.entrySet().forEach(
+                            (e) -> {
+                                // Occurs when a bookie need to be replaced at very start of recovery
+                                if (lastEnsembleKey.get().equals(e.getKey())) {
+                                    builder.replaceEnsembleEntry(e.getKey(), e.getValue());
+                                } else {
+                                    builder.newEnsembleEntry(e.getKey(), e.getValue());
+                                }
+                            });
+                    return builder.closingAt(lac, len).build();
+                },
+                this::setLedgerMetadata).run();
+        f.thenRun(() -> newEnsemblesFromRecovery.clear());
+        return f;
+    }
+
+    @Override
+    List<BookieSocketAddress> getCurrentEnsemble() {
+        if (newEnsemblesFromRecovery.size() > 0) {
```

Comment:
nit: it is typically recommended to use `isEmpty` for size checking rather than calling `size`. because size on some data structures can be an expensive operation.

Reply:
sure, will change.